# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv("/Users/kouamengoran/Desktop/SpringboardF/data/step3_output.csv")

In [3]:
df.head()

Unnamed: 0                           Name       state  summit_elev  \
0           1            Eaglecrest Ski Area      Alaska         2600   
1           2               Hilltop Ski Area      Alaska         2090   
2           4            Sunrise Park Resort     Arizona        11100   
3           5  Yosemite Ski & Snowboard Area  California         7800   
4           8         Boreal Mountain Resort  California         7700   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0           1540      0        0.0          0          0     0  ...   
1            294      0        0.0          0          0     0  ...   
2           1800      0        0.0          0          1     2  ...   
3            600      0        0.0          0          0     0  ...   
4            500      0        0.0          0          1     1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0            60.0              45.0       44.0   
1               30.0            30.0             150.0       36.0   
2              800.0            80.0             115.0       49.0   
3               88.0             0.0             110.0       84.0   
4              380.0           200.0             150.0       54.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0      53.00000               90.0   
1             69.0          30.0      34.00000              152.0   
2            250.0          74.0      78.00000              104.0   
3            300.0          47.0      47.00000              107.0   
4            400.0          49.0      64.16681              150.0   

   NightSkiing_ac  clusters  
0             0.0         2  
1            30.0         1  
2            80.0         0  
3             0.0         0  
4           200.0         0  

[5 rows x 27 columns]

In [4]:
# chech the columns trams , fastEight , tastSixes 
df['trams'].value_counts()

0    157
Name: trams, dtype: int64

In [5]:
df['fastEight'].value_counts()

0.0    157
Name: fastEight, dtype: int64

In [6]:
df['fastSixes'].value_counts()

0    157
Name: fastSixes, dtype: int64

**<font color='teal'>The columns value of the variables trams, fastEight,and fastsixes is `zero`.These varaibles do not provide additional information to the data. We will remove them </font>**

In [7]:
df.drop(['trams', 'fastEight', 'fastSixes', 'Unnamed: 0'], axis=1 , inplace=True)

In [8]:
# check the shape of the dataset 
df.shape 

(157, 23)

In [9]:
# Check the first five  rows 
df.head()

Name       state  summit_elev  vertical_drop  \
0            Eaglecrest Ski Area      Alaska         2600           1540   
1               Hilltop Ski Area      Alaska         2090            294   
2            Sunrise Park Resort     Arizona        11100           1800   
3  Yosemite Ski & Snowboard Area  California         7800            600   
4         Boreal Mountain Resort  California         7700            500   

   fastQuads  quad  triple  double  surface  total_chairs  ...  \
0          0     0       0       4        0             4  ...   
1          0     0       1       0        2             3  ...   
2          1     2       3       1        0             7  ...   
3          0     0       1       3        1             5  ...   
4          1     1       3       1        2             8  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0            60.0              45.0       44.0   
1               30.0            30.0             150.0       36.0   
2              800.0            80.0             115.0       49.0   
3               88.0             0.0             110.0       84.0   
4              380.0           200.0             150.0       54.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0      53.00000               90.0   
1             69.0          30.0      34.00000              152.0   
2            250.0          74.0      78.00000              104.0   
3            300.0          47.0      47.00000              107.0   
4            400.0          49.0      64.16681              150.0   

   NightSkiing_ac  clusters  
0             0.0         2  
1            30.0         1  
2            80.0         0  
3             0.0         0  
4           200.0         0  

[5 rows x 23 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [10]:
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df.state)],  axis=1)

In [11]:
df.shape

(157, 56)

In [12]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Mexico', 'New York', 'North Carolina', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming'],
      dtype='object')

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [13]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler


# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler()

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled= scaler.fit_transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [15]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [16]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred= lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [17]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [18]:
from sklearn.metrics import explained_variance_score, mean_absolute_error
explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred)



(0.753255510586298, 6.071273923087126)

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [19]:
# Print the model  intercept  of the model 
print(lm.intercept_)

8.967259578800814


In [ ]:
# Print the model  coefficient 
print(lm.coef_)

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [21]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        1.590781e-03
vertical_drop      2.047017e-03
fastQuads          5.375312e-01
quad               4.050648e-01
triple             1.231926e+00
double             3.403492e-01
surface            1.113053e+00
total_chairs       7.211189e-01
Runs               5.557695e-02
TerrainParks       3.065000e-01
LongestRun_mi      7.841830e-01
SkiableTerrain_ac  4.209226e-03
Snow Making_ac     1.204565e-03
daysOpenLastYear   1.542289e-02
yearsOpen          7.798191e-02
averageSnowfall    8.728925e-03
AdultWeekday       5.959501e-01
projectedDaysOpen  2.518910e-02
NightSkiing_ac     2.138484e-02
clusters           1.362406e+00
Alaska             1.607460e-01
Arizona            1.284962e+01
California         6.097820e+00
Colorado           9.250828e+00
Connecticut        6.755653e+00
Idaho              1.288981e+01
Illinois           2.796079e+00
Indiana            2.529514e+00
Iowa               1.491120e+00
Maine              1.017173e+01
Maryland           7.993606e-15
Massachusetts      1.272176e+01
Michigan           1.980067e+00
Minnesota          5.979209e-01
Missouri           5.447184e+00
Montana            1.425165e+01
Nevada             1.192371e+01
New Hampshire      6.242096e+00
New Mexico         1.434579e+01
New York           3.990968e+00
North Carolina     1.018724e+01
Ohio               3.998156e+00
Oregon             9.434522e+00
Pennsylvania       7.230367e+00
Rhode Island       8.881784e-16
South Dakota       0.000000e+00
Tennessee          2.040761e+01
Utah               5.659797e+00
Vermont            8.619743e+00
Virginia           1.631477e+01
Washington         4.325937e+00
West Virginia      7.476710e+00
Wisconsin          5.612178e+00
Wyoming            1.338291e+01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

 For this model we will remove all the dummies variables created from the variable state.

In [22]:
#  let's check the five first rows 
df.head()

Name  summit_elev  vertical_drop  fastQuads  quad  \
0            Eaglecrest Ski Area         2600           1540          0     0   
1               Hilltop Ski Area         2090            294          0     0   
2            Sunrise Park Resort        11100           1800          1     2   
3  Yosemite Ski & Snowboard Area         7800            600          0     0   
4         Boreal Mountain Resort         7700            500          1     1   

   triple  double  surface  total_chairs  Runs  ...  Rhode Island  \
0       0       4        0             4  36.0  ...             0   
1       1       0        2             3  13.0  ...             0   
2       3       1        0             7  65.0  ...             0   
3       1       3        1             5  10.0  ...             0   
4       3       1        2             8  33.0  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 56 columns]

In [23]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Mexico', 'New York', 'North Carolina', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming'],
      dtype='object')

In [24]:
 # list of features to drop 
state= ['Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Mexico', 'New York', 'North Carolina', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming']

In [25]:
# let's drop the states from the dataset 
df.drop(state, axis=1 ,inplace = True)

In [26]:
# Let's inspect the shape of the dataset 
df.shape

(157, 22)

In [27]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler()

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled= scaler.fit_transform(X) 

In [28]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [29]:
# let's fit the model 2 
model = lm.fit(X_train,y_train)

In [30]:
# predict the y values of the model
y_pred= lm.predict(X_test)

In [31]:
# The  variance_score and the mean_absolute _error of the model 
explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred)


(0.7640087410862837, 6.143019540487069)

In [32]:
# The y intercept of the model
print(lm.intercept_)  

7.49027007071404


In [33]:
# the coeffient of the differents  features 
print(lm.coef_)

[ 2.81495485e-04  4.26200364e-03 -8.09537602e-01  1.40828032e+00
  6.23682867e-01 -4.21042353e-01 -2.61164902e-01  5.40218330e-01
  2.93608585e-03  7.39754218e-01  2.81457096e-01 -4.11169089e-03
  3.93575017e-03  7.95056318e-03  4.59296051e-02 -8.60762262e-03
  6.67834631e-01  2.70245319e-03  1.31794303e-02  5.03594700e+00]


In [34]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev           0.000281
vertical_drop         0.004262
fastQuads             0.809538
quad                  1.408280
triple                0.623683
double                0.421042
surface               0.261165
total_chairs          0.540218
Runs                  0.002936
TerrainParks          0.739754
LongestRun_mi         0.281457
SkiableTerrain_ac     0.004112
Snow Making_ac        0.003936
daysOpenLastYear      0.007951
yearsOpen             0.045930
averageSnowfall       0.008608
AdultWeekday          0.667835
projectedDaysOpen     0.002702
NightSkiing_ac        0.013179
clusters              5.035947

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [35]:
# drop the variable summit_elev, base_elev
df.drop(['summit_elev'],axis= 1, inplace= True )

In [36]:
df.shape

(157, 21)

In [37]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler()

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled= scaler.fit_transform(X) 

In [38]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [39]:
# fit the model3 
model3 = lm.fit(X_train,y_train)

In [40]:
# predict the y  value of model 3 linear regression 
y_pred= lm.predict(X_test)

In [41]:
# the variance_score and the mean absolute error of the model3
explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred)

(0.7648260895619037, 6.125625632235559)

In [42]:
# the intercept of models linear gression 
print(lm.intercept_)

8.601459020461462


In [43]:
print(lm.coef_)

[ 4.70575161e-03 -7.54512717e-01  1.54843236e+00  6.14212508e-01
 -4.58368467e-01 -3.92915142e-01  5.56848545e-01  4.51038866e-03
  7.63039277e-01  3.99106639e-01 -4.26733869e-03  2.12292996e-03
  5.72773175e-03  4.60279387e-02 -7.06087009e-03  6.72560393e-01
  4.89480466e-03  1.09516175e-02  4.30721121e+00]


In [44]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
vertical_drop         0.004706
fastQuads             0.754513
quad                  1.548432
triple                0.614213
double                0.458368
surface               0.392915
total_chairs          0.556849
Runs                  0.004510
TerrainParks          0.763039
LongestRun_mi         0.399107
SkiableTerrain_ac     0.004267
Snow Making_ac        0.002123
daysOpenLastYear      0.005728
yearsOpen             0.046028
averageSnowfall       0.007061
AdultWeekday          0.672560
projectedDaysOpen     0.004895
NightSkiing_ac        0.010952
clusters              4.307211

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

#### | Model | Explained Variance| Mean Absolute Error|Features Dropped|
| Model 1. | 0.753255510586298, 6.071273923087126  |
| Model 2. | 0.7640087410862837, 6.143019540487069|'state'|
| Model 3. | 0.7648260895619037 , 6.125625632235559|'state','summit_elev , 'base_elev'||

Model Selection:

**<font color='darkcyan'>The state features do not provide insightful information to Big Mountain management, so I eliminate model1. Now, I compare model 2 and model 3 mean absolute error. Model 2 's mean absolute error is higher than Model 3. Therefore, I select model 3  to predict the response variable. </font>**